collectors/proc.plugin/proc_pagetypeinfo.c

```diff
@@ -0,0 +1,344 @@
+        sum += p->free_pages_size[o] * (o+1) * PAGE_SIZE;
+    return sum;
+}
+
+static inline uint64_t pageline_total_count(struct pageline *p) {
+    uint64_t sum = 0, o;
+    for (o=0; o<MAX_PAGETYPE_ORDER; o++)
+        sum += p->free_pages_size[o];
+    return sum;
+}
+
+// Check if a line of /proc/pagetypeinfo is valid to use
+// Free block lines starts by "Node" && 4th col is "type"
+#define pagetypeinfo_line_valid(ff, l) (strncmp(procfile_lineword(ff, l, 0), "Node", 4) == 0 && strncmp(procfile_lineword(ff, l, 4), "type", 4) == 0)
+
+// Dimension name from the order
+#define dim_name(s, o) (snprintfz(s, 16,"%luKB (%lu)", (1 << o) * PAGE_SIZE / 1024, o))
+
+int do_proc_pagetypeinfo(int update_every, usec_t dt) {
+    (void)dt;
+
+    // Config
+    static int do_global, do_detail;
+    static SIMPLE_PATTERN *filter_types = NULL;
+
+    // Counters from parsing the file, that doesn't change after boot
+    static int cnt_nodes = -1, cnt_zones = -1, cnt_pagetypes = -1, cnt_pageorders = -1;
+    static struct systemorder systemorders[MAX_PAGETYPE_ORDER] = {};
+    static struct pageline* pagelines = NULL;
+    static size_t pagelines_cnt = -1, lines = -1;
+
+    // Handle
+    static procfile *ff = NULL;
+
+    // RRD Sets
+    static RRDSET *st_order = NULL;
+    static RRDSET **st_nodezonetype = NULL;
+
+    // Local temp variables
+    size_t l, o, p;
+    struct pageline *pgl = NULL;
+
+    // --------------------------------------------------------------------
+    // Startup: Open /proc/pagetypeinfo
+    if(unlikely(!ff)) {
+        ff = procfile_open(PLUGIN_PROC_MODULE_PAGETYPEINFO_NAME, " \t,", PROCFILE_FLAG_DEFAULT);
+    }
+    if(unlikely(!ff))
+        return 1;
+
+    ff = procfile_readall(ff);
+    if(unlikely(!ff))
+        return 0; // we return 0, so that we will retry to open it next time
+
+    // --------------------------------------------------------------------
+    // Init: find how many cnt_nodes, Zones and Types
+    if(unlikely(cnt_nodes == -1)) {
+        size_t nodenumlast = -1;
+        char *zonenamelast = NULL;
+
+        lines = procfile_lines(ff);
+        if(unlikely(!lines)) {
+            error("PLUGIN: PROC_PAGETYPEINFO: Cannot read %s, zero lines reported.", PLUGIN_PROC_MODULE_PAGETYPEINFO_NAME);
+            return 1;
+        }
+
+        // Configuration
+        do_global = config_get_boolean(CONFIG_SECTION_PLUGIN_PROC_PAGETYPEINFO, "enable system summary", CONFIG_BOOLEAN_YES);
+        do_detail = config_get_boolean_ondemand(CONFIG_SECTION_PLUGIN_PROC_PAGETYPEINFO, "enable detail per-type", CONFIG_BOOLEAN_AUTO);
+        filter_types = simple_pattern_create(
+                config_get(CONFIG_SECTION_PLUGIN_PROC_PAGETYPEINFO, "hide charts id matching", "")
+                , NULL
+                , SIMPLE_PATTERN_SUFFIX
+        );
+
+
+        // 4th line is the "Free pages count...". Just substract the 8 words.
+        cnt_pageorders = procfile_linewords(ff, 3) - 9;
```

Comment:
The 4th line is always the list of orders (that is, constant 
[mm/vmstat.c :: pagetypeinfo_show](https://github.com/torvalds/linux/blob/master/mm/vmstat.c#L1502-L1505) + [mm/vmstat.c :: pagetypeinfo_showfree](https://github.com/torvalds/linux/blob/master/mm/vmstat.c#L1402-L1404) 
Should always be 11, but can be overridden [include/linux/mmzone.h](https://github.com/torvalds/linux/blob/master/include/linux/mmzone.h#L25-L30)

Reply:
What if the user set a wrong file in the configuration and the `cnt_pageorders` becomes negative?

Reply:
It wouldn't have an effect in current code, but it showed me the parsing loops were using the constant `MAX_PAGETYPE_ORDER` insteada of `cnt_pageorders`. 
So I added more strict checks for value, and tried to make it parse a buggy pagetypeinfo file.

Reply:
When unrolling this, seems that if we call `callocz_int` with 0 object, it will `malloc` only 1 `size_t`, then try to return `(void *)&n[1];`, which generates a segfault.